narwhal/executor/src/subscriber.rs

```diff
@@ -55,47 +52,85 @@ struct Fetcher<Network> {
     metrics: Arc<ExecutorMetrics>,
 }
 
-#[must_use]
-pub fn spawn_subscriber(
+pub fn spawn_subscriber<State: ExecutionState + Send + Sync + 'static>(
     name: PublicKey,
     network: oneshot::Receiver<P2pNetwork>,
     worker_cache: SharedWorkerCache,
     committee: Committee,
     rx_reconfigure: watch::Receiver<ReconfigureNotification>,
     rx_sequence: metered_channel::Receiver<CommittedSubDag>,
-    tx_notifier: metered_channel::Sender<(BatchIndex, Batch)>,
     metrics: Arc<ExecutorMetrics>,
     restored_consensus_output: Vec<CommittedSubDag>,
-) -> JoinHandle<()> {
-    spawn_monitored_task!(async move {
-        // This is ugly but has to be done this way for now
-        // Currently network incorporate both server and client side of RPC interface
-        // To construct server side we need to set up routes first, which requires starting Primary
-        // Some cleanup is needed
-        let network = network.await.expect("Failed to receive network");
-        info!("Starting subscriber");
-        let network = SubscriberNetworkImpl {
+    state: State,
+) -> Vec<JoinHandle<()>> {
+    // This is ugly but has to be done this way for now
+    // Currently network incorporate both server and client side of RPC interface
+    // To construct server side we need to set up routes first, which requires starting Primary
+    // Some cleanup is needed
+
+    let (tx_notifier, rx_notifier) =
+        metered_channel::channel(primary::CHANNEL_CAPACITY, &metrics.tx_notifier);
+
+    vec![
+        spawn_monitored_task!(run_notify(state, rx_notifier)),
+        spawn_monitored_task!(create_and_run_subscriber(
             name,
+            network,
             worker_cache,
             committee,
-            network,
-        };
-        let fetcher = Fetcher {
-            network,
-            metrics: metrics.clone(),
-        };
-        let subscriber = Subscriber {
             rx_reconfigure,
             rx_sequence,
             metrics,
+            restored_consensus_output,
             tx_notifier,
-            fetcher,
-        };
-        subscriber
-            .run(restored_consensus_output)
-            .await
-            .expect("Failed to run subscriber")
-    })
+        )),
+    ]
+}
+
+async fn run_notify<State: ExecutionState + Send + Sync + 'static>(
+    state: State,
+    mut tr_notify: metered_channel::Receiver<ConsensusOutput>,
+) {
+    loop {
```

Comment:
Is this `loop` needed here? I believe the `while` on its own should be enough

Reply:
Also I get that this task will leave for as long the `tx/tr_notify` channels will live, but I would still suggest making it listen to the `tx_reconfigure` for any shutdown messages - let's keep things explicit

Reply:
✔️